lib/Sema/CSApply.cpp

```diff
@@ -4615,6 +4617,10 @@ namespace {
         conformances.push_back(hashableConformance);
       }
 
+      auto component =
+          KeyPathExpr::Component::forSubscriptWithPrebuiltIndexExpr(
+              ref, newIndexExpr, newLabels, resolvedTy, componentLoc, {});
```

Comment:
`forSubscriptWithPrebuiltIndexExpr` doesn't appear to make a copy of the argument labels you pass it, so you'll need to make a copy with e.g `cs.getASTContext().AllocateCopy(newLabels)`. It might even be worth performing a copy in the `KeyPathExpr::Component` constructor to defensively avoid future memory bugs here.

Also, if you want, you could pass `conformances` into `forSubscriptWithPrebuiltIndexExpr` here, as long as it also gets copied.

Reply:
Maybe `forSubscriptWithPrebuiltIndexExpr` should be updated to use placement new and return a pointer. As is, it doesn't return a pointer so I don't think we can/should copy `component`.

Reply:
We don't want to be copying `component`, sorry what I meant was we could perform the copying operation within:

https://github.com/apple/swift/blob/cd3ada5abbe78d4062b7d323e5fd15a2df3d3524/lib/AST/Expr.cpp#L2145-L2163

rather than at the call-site. But given we only need to perform the copy at this one call-site, it's probably not worth it for now.